### PR TITLE
Added 40382$ utility for local dom querySelectorAll

### PR DESCRIPTION
**Pull request by joineral32**
Added a $$$ convenience method for querySelectorAll against an element's local dom, similar to $$. See issue: https://github.com/Polymer/polymer/issues/3183

### CODE REVIEW

**Comment by samccone** [comment]
Thanks for the PR @joineral32, someone on the team will take a look at this.

**Comment by dsschnau** [comment]
+1, i was just about to open an issue requesting this sugar

File: test/unit/utils.html [hunk 1]
```diff
@@ -36,6 +46,50 @@
...
+
+      var byId = window.el1.$$('#id-selectable');
+      assert(byId, 'querying local dom by id selector returns element with given id');
+      assert(!window.el1.$$('#id-nosuchthing'), 'querying local dom by id selector returns no element if given id does not exist');
+    });
+
+    test('$$$', function () {
+      var byTag = window.el1.$$$('span');
+      assert(byTag.length === 3
+              && byTag[0].tagName === 'SPAN'
```
Comment (TimvdLippe) [inline test/unit/utils.html]: It is better to split these asserts into seperate asserts. When one of the conditions is false, the error message is more intuitive and easier to debug.

Comment (joineral32-cengage) [inline test/unit/utils.html]: Sure thing! I'm hoping to get around to making that change this weekend.

**Comment by sorvell** [comment]
Sorry for the delay and thanks for the effort here. We think it was probably a mistake to provide public sugaring for querying and element's local dom via `$$`. It's also a poorly named function. Because of these points, we don't want to expand the problem here.

**Comment by joineral32-cengage** [comment]
Certainly we don't want to add this feature if $$ was deemed a poor choice. May I ask what are the factors that have lead you all to deem public $$ to be hazardous? (and maybe add these to the official documentation?) It would be valuable in future design decisions to understand these points.

**Comment by sorvell** [comment]
`$$` queries the element's local dom which is its private implementation. There's not a good way to make this explicitly private and indeed under Shadow DOM one can do `element.shadowRoot.querySelector`. We chose a slightly obscure name to slightly obscure the function. Our current consensus is that it was just a little too unclear.

**Comment by ktiedt** [comment]
@sorvell if the intent was for `$$` to be private or not used... why was it so clearly documented as a Utility function? That seems to imply it was intended to be used... https://www.polymer-project.org/1.0/docs/devguide/utility-functions.html#utility-functions

**Comment by jay8t6** [comment]
are you guys gonna merge this soon or are we still directed to use querySelectorAll? It would be nice if we can use $$$.

**Comment by ktiedt** [comment]
Considering this is marked as closed, and has had no further follow ups since March, I would say its pretty much decided it is not landing.

**Comment by jay8t6** [comment]
Gotcha. So it's not recommended to use this.$$?

**Comment by ktiedt** [comment]
I'm pushing to get more concrete answers on those statements.

**Comment by tjsavage** [comment]
`this.$$` is strictly shorthand for `Polymer.dom(this.root).querySelector()`. The original thought was writing `Polymer.dom(this.root).querySelector()` was too verbose, so it'd be nice to build a relatively obscure API that's shorter but does the same thing. `$$` was chosen to be obscure so that it doesn't seem like "public" API on every element, as the DOM it queries over should be considered private.

With the benefit of hindsight, we've come to the position that using the platform API here makes it easier to figure out what is going on. While `Polymer.dom(this.root).querySelector()` is verbose, it's much more understandable than having to go to the documentation to figure out what `this.$$` means. It tells you exactly what it's doing - using the Shady DOM shim to query select over the element's shadow/shady root.

So in short: we think using `Polymer.dom(this.root).querySelector()` actually makes for better code. Of course, `this.$$` will continue to be supported, as there's plenty of code out there that uses it. In future versions of Polymer, this sort of obscuring API will likely be dropped. We'll look to update the docs to reflect this position.

**Comment by jay8t6** [comment]
Makes sense. Thank you for the explanation.
